backend/datasource/caching.go

```diff
@@ -0,0 +1,179 @@
+package datasource
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+var (
+	// ErrorNotCached is returned when a cached value was not found in the internal cache
+	ErrorNotCached = errors.New("no cached response found")
+)
+
+// A CachedResponse is the combination of a response and an expiration timestamp
+type CachedResponse struct {
+	ExpiresAt time.Time
+	Frames    data.Frames
+}
+
+func getCacheKey(duration time.Duration, req backend.DataQuery) (string, error) {
+	m := map[string]interface{}{
+		"query":    req.JSON,
+		"interval": req.Interval,
+		"type":     req.QueryType,
+		"time_range": backend.TimeRange{
+			To:   req.TimeRange.To.Round(duration),
+			From: req.TimeRange.From.Round(duration),
+		},
+	}
+
+	b := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(b).Encode(m); err != nil {
+		return "", err
+	}
+
+	h := sha256.New()
+	if _, err := h.Write(b.Bytes()); err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(h.Sum(nil)), nil
+
+}
+
+// CachedQueryDatahandler is a QueryDataHandler wrapper that responds with cached values
+type CachedQueryDatahandler struct {
+	duration time.Duration
+	handler  backend.QueryDataHandler
+	cache    map[string]CachedResponse
```

Comment:
This would need to be locked. We probably don't need to implement our own caching, I think we should look at other libraries we could used based on properties we would want from the cache. Consider perhaps single flight (if a identical queries come in, they wait for one to finish and share the results).

Reply:
Good call, I totally forgot.

Maybe you're right but I also wonder if it's worth bringing in a third party library for this since it's intended to be pretty simple. Do you have any in mind?